### PR TITLE
api: adjust the behavior of header's present_match

### LINKED ISSUE #14656: API: router HeaderMatch "present_match" field is a bool but value is ignored
The HeaderMatch [present_match field](https://github.com/envoyproxy/envoy/blob/748a9009e840c4f8da3a4021aa33f81d68712d8f/api/envoy/config/route/v3/route_components.proto#L1857) is a boolean field, but Envoy does not actually use the field's value at all -- if the field is present in the `oneof`, then it [unconditionally matches if the header is present](https://github.com/envoyproxy/envoy/blob/e80b8eec292382f28ac0428229ce165d12ddeb13/source/common/http/header_utility.cc#L159), regardless of the value.

This means that if present_match=false, it matches when the header is present, which is extremely counter-intuitive.  And the behavior is certainly not obvious to someone reading the proto file.  In gRPC, we assumed that because it's a boolean field that the value actually mattered, so we implemented it such that if the value is false, the match is true only if the header is *not* present.  This means that Envoy and gRPC are behaving differently here.

@htuch did some digging for me and found that this field was added in #3443, at the same time as the [invert_match field](https://github.com/envoyproxy/envoy/blob/748a9009e840c4f8da3a4021aa33f81d68712d8f/api/envoy/config/route/v3/route_components.proto#L1891).  Perhaps the thinking was that since there was a separate way to invert any match type, setting present_match (to any value) and invert_match=true is the right way to match on the absence of a field.  However, in that case, it would have made more sense for the present_match field to be of type `google.protobuf.Empty` instead of being a boolean.

I think we need to choose one of the following options:

1. Define the correct behavior as gRPC's behavior.  This is a boolean field, which means that a value of false indicates that the header must be absent (which means that {present_match=false, invert_match=true} is the same as {present_match=true, invert_match=false}).  In this case, we can simply change Envoy's implementation to match gRPC's, and no API changes are needed.

2. Define the correct behavior as Envoy's behavior.  In this case, the API is wrong, and we should deprecate the existing present_match field and replace it with a new one with type `google.protobuf.Empty`, so that the semantics are no longer confusing.

I'd ideally like to go with (1), because it's less work overall and leaves us with less API cruft.  But I don't know what the risk of breaking existing users is, because I don't know if anyone is actually setting this field to false and relying on the existing behavior.

CC @envoyproxy/api-shepherds @voidzcy @menghanl @ashithasantosh

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: api: adjust the behavior of header's present_match

Additional Description:
Currently, envoy ignores the value of present match. This commit adjust
envoy's behavior. When the value is True, envoy will check the header
present. When the value is False, envoy will check the header absent.
Risk Level: low
Testing: unit test added
Docs Changes: api doc updated
Release Notes: minor behavior change
Fixes #14656

Signed-off-by: He Jie Xu <hejie.xu@intel.com>

### CODE REVIEW

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -1885,8 +1885,8 @@ message HeaderMatcher {
     //   "-1somestring"
     type.v3.Int64Range range_match = 6;
 
-    // If specified, header match will be performed based on whether the header is in the
-    // request.
+    // If specified as true, header match will be performed based on whether the header is in the
+    // request. If specified as false, header match will be performed based on whether the header is absent.
```
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:1889]: I think this is technically a data plane behavioral change. That said, I don't see any sensible way that this field could have been configured with `false` and a meaningful match expected previously, so I don't think we need runtime protection.

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -29,11 +29,11 @@ Minor Behavior Changes
   ``envoy.reloadable_features.send_strict_1xx_and_204_response_headers``
   (do not send 1xx or 204 responses with these headers). Both are true by default.
 * http: serve HEAD requests from cache.
+* http: the behavior of the present_match in route header matcher changed. The value of present_match is ignored in the past. The new behavior is present match performed when value is true. absent match performed when the value is false.
```
Comment (htuch) [inline docs/root/version_history/current.rst:32]: Please add a RST link to the `present_match` field.

File: test/common/http/header_utility_test.cc [hunk 3]
```diff
@@ -512,7 +512,24 @@ present_match: true
...
+  )EOF";
+
+  std::vector<HeaderUtility::HeaderDataPtr> header_data;
+  header_data.push_back(
+      std::make_unique<HeaderUtility::HeaderData>(parseHeaderMatcherFromYaml(yaml)));
+  EXPECT_FALSE(HeaderUtility::matchHeaders(matching_headers, header_data));
+  EXPECT_TRUE(HeaderUtility::matchHeaders(unmatching_headers, header_data));
+}
+
+TEST(MatchHeadersTest, HeaderPresentInverseMatchWithTrueValue) {
```
Comment (htuch) [inline test/common/http/header_utility_test.cc:532]: Can you add one liner `//` comments above these test cases explaining what they do? It's kind of hard to infer from name.

**Review by htuch — CHANGES_REQUESTED** [review]
Looks good, thanks for the fix! A few minor comments.
/wait

File: docs/root/version_history/current.rst (hunk 2 above)
Comment (soulxu) [reply docs/root/version_history/current.rst]: done, thanks

File: test/common/http/header_utility_test.cc (hunk 3 above)
Comment (soulxu) [reply test/common/http/header_utility_test.cc:538]: done, thanks

**Comment by soulxu** [comment]
@htuch thanks for the review, addressed your comment.

File: docs/root/version_history/current.rst [hunk 4]
```diff
@@ -29,11 +29,12 @@ Minor Behavior Changes
   ``envoy.reloadable_features.send_strict_1xx_and_204_response_headers``
   (do not send 1xx or 204 responses with these headers). Both are true by default.
 * http: serve HEAD requests from cache.
+* http: the behavior of the present_match in route header matcher changed. The value of present_match is ignored in the past. The new behavior is present match performed when value is true. absent match performed when the value is false. Please reference :ref:`present_match
```
Comment (htuch) [inline docs/root/version_history/current.rst:32]: Nit: `*present_match*` etc. (or RST links everywhere)

Comment (soulxu) [reply docs/root/version_history/current.rst:32]: got it, thanks

**Review by htuch — APPROVED** [review]
LGTM, thanks!

File: api/envoy/config/route/v3/route_components.proto (hunk 1 above)
Comment (hanjm) [reply api/envoy/config/route/v3/route_components.proto:1889]: the doc is not correct, config `present_match: false` will not effected, shoud use 
```
present_match: true
invert_match:true
```

Comment (soulxu) [reply api/envoy/config/route/v3/route_components.proto:1889]: @hanjm I tested it on the main branch, it works as expected. I pasted the detail at https://github.com/envoyproxy/envoy/issues/17345#issuecomment-1177455864

Comment (hanjm) [reply api/envoy/config/route/v3/route_components.proto:1889]: sorry, my version(1.17) is too low. Thanks you reply.

**Pull request merged**